instana/instrumentation/google/cloud/pubsub.py

```diff
@@ -0,0 +1,106 @@
+# (c) Copyright IBM Corp. 2021
+# (c) Copyright Instana Inc. 2021
+
+from __future__ import absolute_import
+
+import json
+import wrapt
+from opentracing import Format
+
+from ....log import logger
+from ....singletons import tracer
+
+try:
+    from google.cloud import pubsub_v1
+
+    logger.debug('Instrumenting Google Cloud Pub/Sub')
+
+
+    def _set_publisher_tags(span, topic_path):
+        span.set_tag('gcps.op', 'publish')
+        # Fully qualified identifier is in the form of
+        # `projects/{project_id}/topic/{topic_name}`
+        project_id, topic_name = topic_path.split('/')[1::2]
+        span.set_tag('gcps.projid', project_id)
+        span.set_tag('gcps.top', topic_name)
+
+
+    def _set_consumer_tags(span, subscription_path):
+        span.set_tag('gcps.op', 'consume')
+        # Fully qualified identifier is in the form of
+        # `projects/{project_id}/subscriptions/{subscription_name}`
+        project_id, subscription_id = subscription_path.split('/')[1::2]
+        span.set_tag('gcps.projid', project_id)
+        span.set_tag('gcps.sub', subscription_id)
+
+
+    @wrapt.patch_function_wrapper('google.cloud.pubsub_v1', 'PublisherClient.publish')
+    def publish_with_instana(wrapped, instance, args, kwargs):
+        """References:
+        - PublisherClient.publish(topic_path, messages, metadata)
+        """
+        # check if active
+        parent_span = tracer.active_span
+
+        # return early if we're not tracing
+        if parent_span is None:
+            return wrapped(*args, **kwargs)
+
+        with tracer.start_active_span('gcps-producer', child_of=parent_span) as scope:
+            # trace continuity, inject to the span context
+            trace_continuity_headers = dict()
+            tracer.inject(scope.span.context, Format.TEXT_MAP, trace_continuity_headers)
+
+            # update the metadata dict with instana trace attributes
+            kwargs['trace_continuity_headers'] = json.dumps(trace_continuity_headers)
```

Comment:
The trace context is expected to be present as key-value pairs inside of message attributes. If we serialize it to JSON, a Node consume for example would not pick it up.